### PR TITLE
stream: fix writing to connecting stream

### PR DESCRIPTION
**Pull request by indutny**
If stream is connecting, writing to it was never actually happening,
because we was returning in `uv__stream_io` right after handling it.

/cc @bnoordhuis

### CODE REVIEW

**Comment by bnoordhuis** [comment]
Writing before having connected is currently only supported on Unices and IIRC is hard to implement on Windows (@piscisaureus knows the details, I remember having discussed that with him.)

If the current behavior is problematic, then maybe we should - for the sake of feature parity - make uv_write() simply return ENOTCONN.

@piscisaureus Thoughts?

File: test/test-tcp-write-and-connect.c [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+
+static void write_cb(uv_write_t* req, int status) {
+  ASSERT(status == 0);
+  write_cb_called++;
+}
+
+
+static uv_buf_t alloc_cb(uv_handle_t* handle, size_t size) {
+  uv_buf_t buf;
+  buf.base = (char*)malloc(size);
```
Comment (bnoordhuis) [inline test/test-tcp-write-and-connect.c]: Superfluous cast.

Comment (bnoordhuis) [inline test/test-tcp-write-and-connect.c]: Also, is it necessary to malloc here?

File: test/test-tcp-write-and-connect.c [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+}
+
+
+static void connection_cb(uv_stream_t* tcp, int status) {
+  int r;
+
+  ASSERT(status == 0);
+
+  r = uv_tcp_init(tcp->loop, &incoming);
+  ASSERT(r == 0);
```
Comment (bnoordhuis) [inline test/test-tcp-write-and-connect.c]: You're being somewhat inconsistent with `ASSERT(0 == command)` vs `r = command;  ASSERT(r == 0)` in this file.  I personally prefer the former, it's a bit shorter.

File: test/test-tcp-write-and-connect.c [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+
+  connection_cb_called++;
+  r = uv_read_start((uv_stream_t*) &incoming, alloc_cb, read_cb);
+  ASSERT(r == 0);
+}
+
+
+static void start_server(void) {
+  struct sockaddr_in addr = uv_ip4_addr("0.0.0.0", TEST_PORT);
```
Comment (bnoordhuis) [inline test/test-tcp-write-and-connect.c]: You might as well put the assignment on a new line.  When you merge it into master, you'll have to update it anyway.

File: src/unix/stream.c [hunk 4]
```diff
@@ -1106,7 +1106,8 @@ static void uv__stream_io(uv_loop_t* loop, uv__io_t* w, unsigned int events) {
 
   if (stream->connect_req) {
     uv__stream_connect(stream);
-    return;
+    if (uv__stream_fd(stream) == -1)
+      return; /* connect error */
```
Comment (bnoordhuis) [inline src/unix/stream.c]: I don't think this is necessarily always correct.  For one, you're assuming that the connect callback immediately closes the connection on error.

Comment (indutny) [inline src/unix/stream.c]: Well, I can rephrase comment, but the logic looks sounds to me here.

File: test/test-tcp-write-and-connect.c (hunk 1 above)
Comment (indutny) [inline test/test-tcp-write-and-connect.c]: Nope, I just copied it from other test.

**Comment by indutny** [comment]
All fixed, please review again.

File: src/unix/stream.c (hunk 4 above)
Comment (bnoordhuis) [inline src/unix/stream.c]: "The logic looks sound to me" doesn't really address the issue I raised... at the very least explain _why_ you think that's okay.

Comment (indutny) [inline src/unix/stream.c]: I don't follow, sorry.

Comment (bnoordhuis) [inline src/unix/stream.c]: Okay, so the issue that I see here is that you expect the user to close the handle immediately on error, else handle->io_watcher.fd wouldn't be -1.

It's not a terrible assumption but I don't think it's always going to be true.  If the handle is still open, execution falls through and what happens next depends entirely on the value of the events argument.

I think the way to address that is by having uv__stream_connect() return a status code and act on that.

**Comment by indutny** [comment]
Pushed follow-up, please take a look.

File: src/unix/stream.c [hunk 5]
```diff
@@ -1105,8 +1105,13 @@ static void uv__stream_io(uv_loop_t* loop, uv__io_t* w, unsigned int events) {
...
 
   if (stream->connect_req) {
-    uv__stream_connect(stream);
-    return;
+    /*
+     * There was either connect error, or user's cb that has
+     * closed the socket. Anyway, no reason to continue here without
+     * fd.
+     */
+    if (uv__stream_connect(stream) != 0 || uv__stream_fd(stream) < 0)
```
Comment (saghul) [inline src/unix/stream.c:1113]: What if uv__stream_connect returns EINPROGRESS?

**Comment by saghul** [comment]
Also, as Ben pointed out, looks like this would fail in Windows: https://github.com/joyent/libuv/blob/master/src/win/stream.c#L150 because UV_HANDLE_WRITABLE is set after connecting: https://github.com/joyent/libuv/blob/master/src/win/tcp.c#L1062

So, shouldn't we try and do the same on both platforms? :-)

**Comment by indutny** [comment]
Well... it seems that windows is not capable of buffering writes in stream.c, perhaps, I can just tweak my app to deal with it :)

**Comment by indutny** [comment]
On other hand it may be a good idea to error on unix as well.

**Comment by saghul** [comment]
Yep, that would be nice. Windows seems to give EPIPE, but I think Ben's
suggestion of using ENOTCON sounds better.

**Pull request closed without merging**